accessEntity/api.go

```diff
@@ -0,0 +1,721 @@
+}
+
+func (a *ApiManager) HandleApiProducts(w http.ResponseWriter, r *http.Request) {
+	a.handelEndpoint(EndpointApiProduct, w, r)
+}
+
+func (a *ApiManager) HandleCompanies(w http.ResponseWriter, r *http.Request) {
+	a.handelEndpoint(EndpointCompany, w, r)
+}
+
+func (a *ApiManager) HandleCompanyDevelopers(w http.ResponseWriter, r *http.Request) {
+	a.handelEndpoint(EndpointCompanyDeveloper, w, r)
+}
+func (a *ApiManager) HandleDevelopers(w http.ResponseWriter, r *http.Request) {
+	a.handelEndpoint(EndpointDeveloper, w, r)
+}
+
+func (a *ApiManager) HandleAppCredentials(w http.ResponseWriter, r *http.Request) {
+	a.handelEndpoint(EndpointAppCredentials, w, r)
+}
+
+func extractIdentifiers(pars map[string][]string) (map[string]string, string, error) {
+	m := make(map[string]string)
+	orgs := pars[IdentifierOrganization]
+	if len(orgs) == 0 {
+		return nil, "", fmt.Errorf("no org specified")
+	}
+	org := orgs[0]
+	for k, v := range pars {
+		k = strings.ToLower(k)
+		if Identifiers[k] {
+			if len(v) == 1 {
+				m[k] = v[0]
+			} else {
+				return nil, org, fmt.Errorf("each identifier must have only 1 value")
+			}
+		}
+	}
+	return m, org, nil
+}
+
+func (a *ApiManager) getCompanyDeveloper(org string, ids map[string]string) (*CompanyDevelopersSuccessResponse, *common.ErrorResponse) {
+	valid, keyVals := parseIdentifiers(EndpointCompanyDeveloper, ids)
+	if !valid {
+		return nil, ErrInvalidPar
+	}
+	priKey, priVal := keyVals[0], keyVals[1]
+
+	devs, err := a.DbMan.GetCompanyDevelopers(org, priKey, priVal, "", "")
+	if err != nil {
+		log.Errorf("getCompanyDeveloper: %v", err)
+		return nil, newDbError(err)
+	}
+
+	if len(devs) == 0 {
+		return nil, ErrNotFound
+	}
+
+	var details []*CompanyDeveloperDetails
+	for _, dev := range devs {
+		comName, err := a.DbMan.GetComNames(dev.CompanyId, TypeCompany)
+		if err != nil || len(comName) == 0 {
+			log.Errorf("getCompanyDeveloper: %v", err)
+			return nil, newDbError(err)
+		}
+		email, err := a.DbMan.GetDevEmailByDevId(dev.DeveloperId)
+		if err != nil {
+			log.Errorf("getCompanyDeveloper: %v", err)
+			return nil, newDbError(err)
+		}
+		detail := makeComDevDetails(&dev, comName[0], email)
+		details = append(details, detail)
+	}
+	return &CompanyDevelopersSuccessResponse{
+		CompanyDevelopers:      details,
+		Organization:           org,
+		PrimaryIdentifierType:  priKey,
+		PrimaryIdentifierValue: priVal,
+	}, nil
+}
+
+func (a *ApiManager) getDeveloper(org string, ids map[string]string) (*DeveloperSuccessResponse, *common.ErrorResponse) {
+	valid, keyVals := parseIdentifiers(EndpointDeveloper, ids)
+	if !valid {
+		return nil, ErrInvalidPar
+	}
+	priKey, priVal := keyVals[0], keyVals[1]
+
+	devs, err := a.DbMan.GetDevelopers(org, priKey, priVal, "", "")
+	if err != nil {
+		log.Errorf("getDeveloper: %v", err)
+		return nil, newDbError(err)
+	}
+
+	if len(devs) == 0 {
+		return nil, ErrNotFound
+	}
+	dev := &devs[0]
+
+	attrs := a.DbMan.GetKmsAttributes(dev.TenantId, dev.Id)[dev.Id]
```

Comment:
Just a generic comment about this pattern. I recollect that not all attributes are mandatory. Did you ensure that you error out of proceeding only if the mandatory parameters are missing. Also, the test cases should match the same.

Reply:
Yes. The returned map of `GetKmsAttributes` can be empty, and attrs will be empty slice in this case.

Reply:
ack